index.ts

```diff
@@ -1,35 +1,65 @@
 export class OatyArray {
-  private _original: object[]
-  private _keys: string[]
-  private _transposed: object = {}
+  private keys: string[] | undefined
+  private original: object[] = []
+  private transposed: { [key: string]: any } = {}
+
+  constructor(options?: { data?: object[], keys?: string[] }) {
+    if (options !== undefined) {
+      if (options.data !== undefined) { this.push(...options.data) }
+      if (options.keys !== undefined) { this.keys = options.keys }
+    }
+  }
 
   get length(): number {
-    return this._original.length
+    return this.original.length
   }
 
-  constructor(input: object[], keys: string[]) {
-    this._original = input.map((x) => Object.assign({}, x))
-    this._keys = keys
+  public getKeys(): string[] | undefined {
+    return this.keys
+  }
 
-    this._transposed = transpose(this._keys, this._original)
+  public getOriginal(): object[] {
+    return this.original
   }
 
-  public get(keyName: string, keyValue: string): object[] {
-    return this._transposed[keyName][keyValue]
+  public getTransposed(): object[] {
+    return this.original
   }
 
-  public push(...items: object[]): number {
-    this._transposed = transpose(this._keys, items, this._transposed)
-    return this._original.push(...items)
+  public get(keyName: string, keyValue: string): object[] | undefined {
+    if (this.transposed === undefined

```

Comment:
loads of undefined checks?

Reply:
Same as another comment: If options.keys is defined then we want to transpose only those keys, otherwise we'll use the object's.

But we do need to decide if we want to return something even if it's just an empty array; would make it more consistent and look like it actually tried to search.

Reply:
not sure why `.get` needed the rewrite? imo less readable now

---

> make it look like it actually tried to search

I feel like this is just misleading the consumer of the library - would prefer we returned `undefined`, matching the behaviour of `Array.prototype.find`

Reply:
`.get` needs the undefined checks.
Surely `return this._transposed[keyName][keyValue]` would've thrown an NPE if nothing had been pushed or if that key hadn't been transposed?

Reply:
`find` also returns a singular object for 1 results rather than array of 1 object. 
For me at least, consistently returning an array is easier to handle.

Reply:
Oh wait I forgot `_this.transposed` is initialised to {} . I'll remove the first undefined check

Reply:
> `find` also returns a singular object for when there's 1 result rather than an array of 1 object.

that's because it searches for the first match, rather than all matches

Reply:
oh really? well, that sucks. Soooo, am I returning undefined or [], or is this a new option?

Reply:
return matching object array directly (effectively return `undefined` if there's no match)

Reply:
if we also added a function similar to `Array.prototype.filter` in the future, that'd return `[]` if no matches

Reply:
Hmmm, would it not make more sense to return `[]` for consistency? Our `get` and `Array.find` are very different, and given we're not calling it `find` I don't think there's any risk of someone expecting `undefined`

Reply:
If I was querying data that doesn't exist, `undefined` is exactly what I'd expect

Reply:
Okay, I'll leave it as undefined. I will make it configurable ([], undefined, or throw) in another branch.

Reply:
Resolved. Will do as another feature.